### ISSUE TITLE
Support updating Alias version

### ISSUE THREAD

**Issue by pdkn** [HyperBrain/serverless-aws-alias#64]
Would be wonderful to see support for alias promotion - http://docs.aws.amazon.com/lambda/latest/dg/API_UpdateAlias.html

**Comment by HyperBrain** [comment]
This is how I would see the feature implemented and working

## Purpose

In real-world scenarios alias deployments can be used to represent
different quality stages that are deployed in parallel.

I will use colors to show example aliases here, to avoid misunderstandings
with environments (like prod/dev) and to show it in an universal meaning.

The master alias (e.g. blue) is normally used for any references from the 
outside to represent the currently active and stable deployment. The next
(unstable) release is tested and deployed using a different alias (e.g. green).

So let us have the following deployments (for reasons of simplicity only one
target function is shown here as an example with some randomly selected versions):

```
Customer -> API (blue) -> MyFunc:blue -> MyFunc:10
QA -> API (green) -> MyFunc:green -> MyFunc:15
Developer 1 -> API (dev1) -> MyFunc:dev1 -> MyFunc:21
Developer 2 -> API (dev2) -> MyFunc:dev2 -> MyFunc:23
(current stable development) -> API (dev) -> MyFunc:dev -> MyFunc:25
```

That is, the customer always ends up at the blue deployed alias (the API URL is most
likely hardcoded in the client). The next release is deployed to "green" and the QA
uses that to test the upcoming version. Developers work on their "private" deployments
and the current stable development branch is deployed to "dev".

This setup is assumed for the following sample use cases.

## Sample use cases

### Publish a tested/qualified version

As soon as the QA tests have been finished on the green alias, it should be deployed.
This simply means that Customers should end up at the already tested versions from the
green deployment instead of the blue versions. The alias promotion in this case would 
set the blue alias to the versions of the green deployment, without redeploying anything:

```
Customer -> API (blue) -> MyFunc:blue -> MyFunc:15
QA -> API (green) -> MyFunc:green -> MyFunc:15
```

Now the QA can start testing the current stable development version which will be the next release.
So the green alias is just promoted to the current dev alias.

```
QA -> API (green) -> MyFunc:green -> MyFunc:25
(current stable development) -> API (dev) -> MyFunc:dev -> MyFunc:25
```

The dev alias can then be deployed again with newer versions while green still references MyFunc:25.

This is all done without deploying anything again to AWS. It would just be a redeclaration of an alias.
This ensures that already deployed code is used and there is absolutely no risk regarding quality.
Of course any alias that is only changed via promotion has to be deployed one time to make it available.
If any lambda version is not referenced by any alias anymore, it will automatically be deleted.

## Implementation

### Feasibility

The promotion feature fits perfectly in the alias plugin's architecture and the architecture of the created
alias stacks. All deployed function version CF resources reside in the alias stacks, so a promotion would
mean to only change the CF stack of the promoted alias.
All version CF resources have to be replaced with the ones of the target alias, and the alias CF resources
contained in the alias stack have to be linked to these versions. No logical ids have to be changed and it is
a plain copy, because logical ids must only be unique within a stack.

### CLI

From a CLI perspective the promotion would be a new subcommand that specifies the target alias to which the
selected alias will be promoted to.

```
serverless alias promote --target=myTarget    // Will promote (move) the master alias to myTarget
serverless alias promote --alias=green --target=myTarget    // Will promote (move) green to myTarget
```

## Caveats

There might be differences in the API endpoints that are deployed in the two participating aliases.
The plugin should take this into account and make sure that the API contains the same API configuration
in the promoted alias as the target alias after the promotion.

Although the promotion is purely done online within the AWS infrastructure, there might be the need to record
promotions somewhere in the serverless deployment bucket. This can be decided later as soon as the first
implementation of the feature is ready, but it must not be forgotten.

**Comment by pdkn** [comment]
Amazing - that looks spot on to me and just what we could use. Thanks so much for taking the time to plan this out!

**Comment by Stefan-Aurea** [comment]
This is an excellent feature which I for one would use, as it enables a 'same tested binaries' approach and ensures that versions are linear in time and not duplicated binary-wise.

ex. Version 10 would always be older code than Version 11.

I would also add a point for discussion here.

Technically, if you switch to your newly-created release branch and execute `sls deploy --alias=releasex`, the hashes for all functions should match the hashes for functions deployed to the alias used for the `develop` branch.

So would it not make more sense for this command to be an auto-promotion? (even if `deploy` was specified).

I would argue that this should be the default behaviour, with the requirement of `--force` should you specifically want to enforce deploying new binaries.

**Comment by cgrimal** [comment]
An excellent feature indeed! :+1:

I am testing this plugin at the moment, as the default behavior of serverless regarding stages/aliases does not fit with how AWS has already been used in my team.

That said, today I think it is possible to use aws-cli to do that. For instance (using your example):
```sh
aws lambda update-alias --function-name myFunc --function-version 15 --name blue
```

What do you think?

Additionally, there is a new feature regarding aliases that I am very keen on using: [traffic shifting](https://docs.aws.amazon.com/lambda/latest/dg/lambda-traffic-shifting-using-aliases.html)

Using your example again, it would work like this:
```sh
aws lambda update-alias --function-name myFunc --function-version 15 --name blue --routing-config AdditionalVersionWeights={"10"=0.8}
```
Enabling you to switch only 20% of your traffic to the new version.

But it doesn't work :sad: because of execution role being different from one alias to another:
```
An error occurred (InvalidParameterValueException) when calling the UpdateAlias operation: Specified versions must be associated with the same execution role
```

So when/if this promotion feature is implemented, it would be great to support this feature, but in the meantime, do you think there is any workaround regarding this execution role issue?

Thanks for your help!

*(I may create a PR but as I am quite new to this serverless environment, I'm not sure I'll manage to implement this one)*

**Comment by HyperBrain** [comment]
Hi @cgrimal . Thanks for the ideas.

Your approach with the aws cli sounds quite good as an intermediate workaround. For the roles, you might be able to enforce that the same role is used for multiple stages/aliases, if you define a service role in your `serverless.yml` (provider: role: XXXXX) and setup the role in the resources section.
However, this might need some experiments, because I never tried anything in that direction 😄

**Comment by cgrimal** [comment]
Thanks @HyperBrain for your answer. It really helped me! :smiley: 
I managed to use traffic shifting by enforcing the role used for multiple aliases through
```yml
provider:
  [...]
  role: arn:aws:iam::XXXX:role/my-service-role
```

I did not setup the role in the resources section though, but I think this is because the role already existed and had the correct rights. Is it correct?

Should I add a section like the following:
```yml
resources:
  Resources:
    IamRoleLambdaExecution:
      Type: AWS::IAM::Role
      Properties:
        Path: role/ 
        RoleName: my-service-role
```

**Comment by HyperBrain** [comment]
No, if the role already exists, it should be sufficient to just reference it from the provider section.
